### PR TITLE
Adding template GCStoGCS using Spark SQL

### PR DESCRIPTION
**Pull request by varunika**
Tested for all combinations of input and output format.

### CODE REVIEW

File: python/dataproc_templates/gcs/gcs_to_gcs.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                .option(constants.HEADER, True) \
+                .option(constants.INFER_SCHEMA, True) \
+                .load(gcs_input_location)
+        elif gcs_input_format == constants.FORMAT_JSON:
+            input_data = spark.read \
+                .json(gcs_input_location)
+        
+        if ((gcs_temp_view != "") & (sql_query != "")):
+            # Create temp view on source data
+            input_data.createOrReplaceTempView(gcs_temp_view)
```
Comment (snrssc) [inline python/dataproc_templates/gcs/gcs_to_gcs.py:169]: What is the use of registering the table as temp view here?

**Review by snrssc — COMMENTED** [review]
LGTM , Except one comment for registering dataset as temp view and validation of temp view name as mandatory.

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 1 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:169]: The idea is that if the user wishes to just change the file formats without any transformations, they would not have to provide a query or a temp view name and can easily go ahead with the file format conversion using the same template

Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:169]: @snrssc , @shashank-google are any changes required? or are we good to close this?

Comment (snrssc) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:169]: Hi @varunika  , Two options i see here,

1. We can make it as optional property and use a default value if nothing specified
2.  Always use default value as the tempview name is not to be accessible outside of the program and in downstream processing.

Comment (snrssc) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:169]: Discussed in a call, user has to specify same table name in both tempview and the query provided. For now this template is designed only for single GCS file read and write.  @varunika  to add comments for specifying tempview has to match with tablename used in the query.

Comment (snrssc) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:169]: Thanks @varunika . LGTM

**Review by snrssc — APPROVED** [review]

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 1 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:169]: @snrssc Added new commit to accommodate the requested change. Please review.

File: python/dataproc_templates/gcs/gcs_to_gcs.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Optional, Any
+from logging import Logger
+import argparse
+import pprint
+
+from pyspark.sql import SparkSession, DataFrame, DataFrameWriter
+from google.cloud import storage
```
Comment (NiloFreitas) [inline python/dataproc_templates/gcs/gcs_to_gcs.py:21]: Remove unused dependency

File: python/dataproc_templates/gcs/gcs_to_gcs.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            dest=constants.GCS_TO_GCS_OUTPUT_LOCATION,
+            required=True,
+            help=(
+                'destination GCS location'
+            )
+        )
+        
+        known_args: argparse.Namespace
+        known_args, _ = parser.parse_known_args(args)
+
```
Comment (NiloFreitas) [inline python/dataproc_templates/gcs/gcs_to_gcs.py:126]: Add the argument validation here:

```
if getattr(known_args, constants.GCS_TO_GCS_SQL_QUERY) and not getattr(known_args, constants.GCS_TO_GCS_TEMP_VIEW_NAME):
            sys.exit('ArgumentParser Error: Temp view name cannot be null if you want to do data transformations with query')
```

File: python/dataproc_templates/gcs/gcs_to_gcs.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+            input_data = spark.read \
+                .format(constants.FORMAT_CSV) \
+                .option(constants.HEADER, True) \
+                .option(constants.INFER_SCHEMA, True) \
+                .load(gcs_input_location)
+        elif gcs_input_format == constants.FORMAT_JSON:
+            input_data = spark.read \
+                .json(gcs_input_location)
+        
+        if ((gcs_temp_view != "") & (sql_query != "")):
```
Comment (NiloFreitas) [inline python/dataproc_templates/gcs/gcs_to_gcs.py:167]: After you add the validation in the parse_args method (my other comment), here you can replace line 167 to 176 with the following.

```
        if sql_query:
            # Create temp view on source data
            input_data.createOrReplaceTempView(gcs_temp_view)
            # Execute SQL
            output_data = spark.sql(sql_query)
        else:
            output_data = input_data
```

File: python/dataproc_templates/gcs/gcs_to_gcs.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            
+            # Execute SQL
+            output_data = spark.sql(sql_query)
+            
+        elif (gcs_temp_view == "" & sql_query != ""):
+            logger.error("Temp view name cannot be null if you want to do data transformations with query")
+            exit (1)
+        
+        # Write
+        if (output_partition_column != ""):
```
Comment (NiloFreitas) [inline python/dataproc_templates/gcs/gcs_to_gcs.py:179]: Replace to:
`if output_partition_column:`

File: python/dataproc_templates/gcs/gcs_to_gcs.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, Sequence, Optional, Any
+from logging import Logger
+import argparse
+import pprint
+
```
Comment (NiloFreitas) [inline python/dataproc_templates/gcs/gcs_to_gcs.py:19]: Add **import sys**, since I used **sys** in one of my comments

**Review by NiloFreitas — CHANGES_REQUESTED** [review]
Hi Varunika!
Great great work!!! Sorry I was a bit picky in my comments, but I added a suggestion to simplify the argument parsing related to the sql query and the temp view name.

File: python/dataproc_templates/gcs/README.md [hunk 7]
```diff
@@ -414,3 +414,88 @@ export JARS="gs://spark-lib/bigquery/spark-bigquery-latest_2.12.jar"
...
+## Example submission
+
+```
+export GCP_PROJECT=<project_id>
+export REGION=<region>
+export GCS_STAGING_LOCATION=<gcs-staging-bucket-folder> 
+export JARS=<gcs-bucket-location-containing-jar-file> 
+
+./bin/start.sh \
+-- --template=GCSTOGCS
```
[the remark this answers is not in the document]
Comment (varunika) [reply python/dataproc_templates/gcs/README.md]: Changes made accordingly.

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 2 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py]: Unused dependency removed.

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 3 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:129]: Nice Suggestion. Added the same to the code. Please review.

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 4 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py]: Changes made accordingly.

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 5 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py]: Done

File: python/dataproc_templates/gcs/gcs_to_gcs.py (hunk 6 above)
Comment (varunika) [reply python/dataproc_templates/gcs/gcs_to_gcs.py:20]: Added sys package

**Review by NiloFreitas — APPROVED** [review]
LGTM

**Pull request merged**